src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs

```diff
@@ -0,0 +1,312 @@
+                    return false;
+                }
+
+                for (int i = 0; i < thisType.Instantiation.Length; i++)
+                {
+                    if (!thisType.Instantiation[i].IsEquivalentTo(otherType.Instantiation[i], visited))
+                    {
+                        return false;
+                    }
+                }
+
+                // Generic equivalence only allows the instantiation to be non-equal
+                return thisType.HasSameTypeDefinition(otherType);
+            }
+
+            return IsEquivalentTo_TypeDefinition((MetadataType)thisType.GetTypeDefinition(), (MetadataType)otherType.GetTypeDefinition(), visited);
+
+            static bool IsEquivalentTo_TypeDefinition(MetadataType type1, MetadataType type2, StackOverflowProtect visited)
+            {
+                Debug.Assert(type1.GetTypeDefinition() == type1);
+                Debug.Assert(type2.GetTypeDefinition() == type2);
+
+                var stackOverflowProtectKey = new CastingPair(type1, type2);
+                if (visited != null)
+                {
+                    if (visited.Contains(stackOverflowProtectKey))
+                    {
+                        // we are in the process of comparing these tokens already. Assume success
+                        return true;
+                    }
+                }
+
+                StackOverflowProtect protect = new StackOverflowProtect(stackOverflowProtectKey, visited);
+
+                TypeIdentifierData data1 = type1.TypeIdentifierData;
+                TypeIdentifierData data2 = type2.TypeIdentifierData;
+                if (data1 == null || data2 == null)
+                {
+                    return false;
+                }
+
+                // Check to ensure that the types are actually opted into equivalence
+                if (!type1.IsTypeDefEquivalent || !type2.IsTypeDefEquivalent)
+                    return false;
+
+                if (!data1.Equals(data2))
+                    return false;
+
+                if (type1.Name != type2.Name)
+                    return false;
+
+                if (type1.Namespace != type2.Namespace)
+                    return false;
+
+                var containingType1 = (MetadataType)type1.ContainingType;
+                var containingType2 = (MetadataType)type2.ContainingType;
+
+                // Types must be either not nested, or nested in equivalent types
+                if ((containingType1 == null) != (containingType2 == null))
+                {
+                    return false;
+                }
+
+                if ((containingType1 != null) && !IsEquivalentTo_TypeDefinition(containingType1, containingType2, visited))
+                {
+                    return false;
+                }
+
+                if (type1.IsInterface != type2.IsInterface)
+                    return false;
+                if (type1.IsInterface)
+                {
+                    if (!type2.IsInterface)
```

Comment:
Nit - I believe this check is superfluous; you have already made sure at line 132 that <code>type1.IsInterface</code> is the same as <code>type2.IsInterface</code> so it seems to me that you should be able to safely <code>return true</code> in this conditional statement.

Reply:
Thank you